### ISSUE TITLE
HabPanelViewer doesn't connect

### ISSUE THREAD

**Issue by Webfeger** [vbier/habpanelviewer#17]
## Expected Behavior
HPV does't discover HabPanel Server on Raspberry.
After standby HPV shows a white site with connection issues. and doesn't reaload the site automaticially

HabPanel in Chrome works fine

## Your Environment

* Device: Samsung Galaxy Tab A 2016
* Android Version: 7.0 - Kernel 3.18.14-13041686
* HabPanelViewer Version 2.2.0 Release Build

**Comment by vbier** [comment]
Can you please attach an adb log? See this page for details on how to do this:
https://android.stackexchange.com/questions/33216/how-can-i-get-a-logcat

**Comment by Webfeger** [comment]
[logcat.log](https://github.com/vbier/habpanelviewer/files/1886158/logcat.log)

Hello,
here the required adb log

**Comment by vbier** [comment]
I just had a look at the log. There should be a message "EventSource connection initiated" when the app tried to open the connection to the server. The message is not in the log, instead there is the message "EventSource connection skipped", which should only be logged if the server URL is empty or the connection already has been established.

It might be possible that the app has created an invalid connection object without logging, as there are some code paths where nothing had been logged at all. I have now added logging to these paths to see if this is the case.

Can you please install the updated apk and generate another log?
https://github.com/vbier/habpanelviewer/raw/issue17/hpv-debug.apk

**Comment by Webfeger** [comment]
[logcat2.zip](https://github.com/vbier/habpanelviewer/files/1886572/logcat2.zip)
next try with zipped version

**Comment by vbier** [comment]
It does not connect because you have not defined a single item in the preferences. Can you please define e.g. the command item and see if you get a connection?

**Comment by Webfeger** [comment]
Interesting.. after i choose a temperature item, "not connected" state changed to OpenHab Server URL.

I enabled the device admin and lock screen checkbox. additionaly disabled Habpanelviewer from energy monitoring. where i can allow it to run while tablet is off?

after tablet turn into standby it issn't able to ping this device after few minutes. so wlan will be switch into standby too. 

i testing it again.

**Comment by vbier** [comment]
Energy monitoring is what I meant. Have you set the WIFI to not disconnect in stand-bye? You should be able to see in your router logs whether the tablet disconnects in stand-bye. In order to see if the app is still running, you can enable the "Connected Indicator" (Verbindungsindikator), which makes the app update the defined openHAB item every minute with the current time stamp.

**Comment by vbier** [comment]
I'm not sure this will change anything. There has to be some detail we missed so far. I will add some logging to the code that generates the white page with the error message tomorrow.

**Comment by Webfeger** [comment]
Okay, now i configured my tablet again and use the 5Ghz Wlan AP instead. But still the same issue

**Comment by vbier** [comment]
I have added more logging. Can you please generate another log?

Start HPV, let it go into stand-bye, wait for 20 seconds and turn the tablet back on. Wait 10 seconds, (and make note of the current time so it is easier for me to find the correct spot in the log) and generate the log.

**Comment by Webfeger** [comment]
[logcat3.zip](https://github.com/vbier/habpanelviewer/files/1887615/logcat3.zip)

next try with follow time notes:
17:43 - star hpv
17:44 - standby
17:45 - press home button to wake up
17:45 - exit logging

**Comment by Webfeger** [comment]
https://android.stackexchange.com/questions/20726/galaxy-tab-10-1-doesnt-keep-wifi-connected-despite-settings

Maybe this will be the reason too.. i try to change network/ scan settings

BtW: i tested it yesterday with public release version of HPV and this morning, app habpanel was showing but freezed

**Comment by Webfeger** [comment]
Yes, i see the white screen when the table came out of standby. I already disable the energy control completly, changed from dhcp to static.. but nothing helps.
If i use Google Chrome Webapp, everything works fine after turn tablet on from standby.

**Comment by vbier** [comment]
I don't understand how this is possible. The only thing I can think of is that the error page is shown as a result of the browsers back action, but even then the log should show at least one time that the error message has been generated, which it does not. I will add some more living later.

**Comment by vbier** [comment]
@webfeger, any news?

**Comment by Webfeger** [comment]
Actually HPV hangs after beeing in standby for a while. It's required to close and restart the app.
It freezes with habpanel now, but don't know why. maybe it will happen after using sensor function and connection indicator.

I already disabled all energyfunctions and changed from dhcp to static. But same issues still exists.

Since yesterday i'm using Google Chrome Webapp without any issues. 
I use Tasker for sending tablet battery status too habpenel. It works fine too.

Maybe this will be an generally WIFI Issue on Android 7.0. Android 8 will released soon. maybe this will helps.

**Comment by vbier** [comment]
If you are no longer interested in getting this fixed i will close the issue.

**Comment by Webfeger** [comment]
I'm interested to find a solution, but i don't know whats the reason for this issue.
I compared it again with Chrome App to get more information. but everything works fine. and i can't detect any wifi issues.

But i don't want you to spend your time with this issue, so i will try this app again if Android 8 is released. 
Otherwise if you have some ideas for debugging, i will check it out.

Thanks for fast response and help

**Comment by vbier** [comment]
If I feel like I am wasting my time I will stop to spend it. I would like to find out what the problem is. From the last log it looks like the app is connected all the time. Can you generate another log with the new version?

**Comment by Webfeger** [comment]
I understand that you don't want to waste your time. But i still doesn't know whats the reason for this issue.
Does the new version have the same URL? https://github.com/vbier/habpanelviewer/raw/issue17/hpv-debug.apk ?

**Comment by vbier** [comment]
Yes, the URL for the apk always stays the same.

**Comment by Webfeger** [comment]
[logcat4.zip](https://github.com/vbier/habpanelviewer/files/1908327/logcat4.zip)
Here are the next logfile with follow steps:

19:21  starting hpv
19:21  display turn off /standby
19:22  home button pressed / display turn on
19:23  display turn off again

**Comment by Webfeger** [comment]
Does this Log is long enough for required informations? otherwise i need to logging for few mins/hours to run into issue with blank white page

**Comment by vbier** [comment]
The log has to cover the time when you get the white page with the not connected message, as otherwise I can not see the cause.

**Comment by vbier** [comment]
How long does it take from stand-bye until you get the message? Is it happening consistently?

**Comment by Webfeger** [comment]
Okay, i try to log it few hours and reduce the log with required infos

**Comment by Webfeger** [comment]
There is an option to create a adb log without using usb connection to pc?
there was no issue the whole day.. maybe the power connection will be the reason

**Comment by Webfeger** [comment]
After monitoring a whole day, i disconnected the usb cable this morning because there was no issue. One minute after disconnecting i got the blank page.. i need to create a adb log on local disk instead usb. but don‘t know if its possible without rooting the device

**Comment by vbier** [comment]
The application logs to the android system log while it is running. adb logcat just prints out the system log.

**Comment by Webfeger** [comment]
[logcat8.zip](https://github.com/vbier/habpanelviewer/files/1913031/logcat8.zip)
Here a new log. Connection was lost between 17:00 and 17:50

**Comment by vbier** [comment]
Android returns BLOCKED for the network state when you turn the tablet back on 17:52:56. Have you allowed background data usage for the app in the android settings under "Apps-HPV-Datennutzung"?

**Comment by Webfeger** [comment]
Yes, background data usage is enabled. only the next option: sava data, is disabled.

**Comment by vbier** [comment]
I previously had a listener for network connectivity change which was not notified at all when the connection state changed to BLOCKING. 
Reading up on the internet several people reported that BLOCKING indicates a state where network access seems to be restricted (by energy saving mechanisms), but in which network access works nevertheless.

I have now therefore removed the check that is performed on wake-up and rely only on the state of the listener. This should make HPV try to reload the start page even when the connection is reported as BLOCKING. Please test if this fixes your problems. The apk is updated.

**Comment by Webfeger** [comment]
Its interesting. in other words: wifi is already set to always on, but diditn work correctly. Thats why i cant ping the tablet after few minutes. But Tablet is listed on my AP as connected. 
I reinstalled the new apk and testing it now.

**Comment by Webfeger** [comment]
Now it looks like it works. After turn on Tablet from Standby i see the red reconnect info at bottom left for 1-2 seconds.
I will check it out 2-3 times again

**Comment by Webfeger** [comment]
Now it works great!! 

Sometimes reconnecting line showing up to 4-5 seconds but it works.
Thank you very much!!

**Comment by Webfeger** [comment]
Now i detected an other issue. Not sure if this is related to the latest change.
After turn on from standby hapbanel state are not actual. Refresh Icon doesnt work. Only the Switch Size Button refreshs the actual state. So i think auto refresh didn‘t work correctly

**Comment by vbier** [comment]
The app does no longer try to reload the page when the tablet is turned on, it relies on HABPanel to reload and refresh the items. HABPanel should detect the connection loss (and also when connection is back, as it tries every 5 seconds to connect), and reload all items.

Or are you talking about values in the system information?

Is it reproducible? Does is work in chrome?

**Comment by Webfeger** [comment]
I mean the state of items. If Tablet went to standby and i changed the state from an openhab app of an other device, HPV still shows the old state after turn on from standby. Refresh Icon have no function. Only when i press the resize icon, all states will be changed to the latest state.

HP works in chrome webapp without any issues. But Item buttons are more clearlier in HPV (no sliding, works faster, etc.)

Not sure if it is reproducable on your devices. maybe it will work if you turn off wifi ap and on again while tablet is in standby.

**Comment by vbier** [comment]
I can not reproduce it. I have the tablet in stand-bye, disconnect internet (by removing the network cable from my wireless ap), change an item from my pc, connect internet (plug cable back in), turn tablet on. It directly shows the new value, you can't even see it had an old value.

You should generate another logfile, maybe there is something in it that helps.

**Comment by Webfeger** [comment]
Hm thats strange, i tested it again between 07:00 and 08:00, after wake up from standby reconnecting needs very long and didnt refresh the state. After plugged usb to my pc, hpv refreshing the states. including the time at top right.

i removed all unrelated logs before 06:00 am

[logcat9.zip](https://github.com/vbier/habpanelviewer/files/1922939/logcat9.zip)

**Comment by vbier** [comment]
I just had a look at the log, but I can not see anything problematic in it:

Screen turned on:
```
04-18 07:33:40.059 30572 30572 D HPV-ScreenControllingAc: onStart: set keep on: false
04-18 07:33:40.059 30572 30572 D HPV-ServerConnection: subscribing items: []
04-18 07:33:40.060 30572 30572 D HPV-ServerConnection: subscribing items: []
04-18 07:33:40.060 30572 30572 D HPV-ServerConnection: subscribing command items: []
04-18 07:33:40.060 30572 30572 D HPV-ServerConnection: subscribing items: [, , ]
04-18 07:33:40.060 30572 30572 D HPV-ServerConnection: subscribing items: [TabletTimeStamp]
04-18 07:33:40.060 30572 30572 D HPV-ClientWebView: updateFromPreferences
04-18 07:33:40.060 30572 30572 D HPV-ClientWebView: updateFromPreferences: connected
```

HPV detecting network is back, sending updates:
```
04-18 07:33:43.062 30572 31036 V HPV-ServerConnection: SSE onLogMessage: reconnect: in timer
04-18 07:33:43.065 30572 31036 V HPV-ServerConnection: SSE onLogMessage: reconnect: in timer, connect invoked
04-18 07:33:43.076 30572 31117 V HPV-ServerConnection: SSE onLogMessage: reconnect: in ChannelFutureListener
04-18 07:33:43.076 30572 31117 V HPV-ServerConnection: SSE onLogMessage: reconnect: success in ChannelFutureListener
04-18 07:33:43.077 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelActive
04-18 07:33:43.142 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = null, headerDone=false, line=HTTP/1.1 200 OK
04-18 07:33:43.144 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = 200, headerDone=false, line=X-Accel-Buffering: no
04-18 07:33:43.145 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = 200, headerDone=false, line=Content-Type: text/event-stream
04-18 07:33:43.147 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = 200, headerDone=false, line=Transfer-Encoding: chunked
04-18 07:33:43.149 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = 200, headerDone=false, line=Server: Jetty(9.3.22.v20171030)
04-18 07:33:43.150 30572 31117 V HPV-ServerConnection: SSE onLogMessage: channelRead0: status = 200, headerDone=false, line=
04-18 07:33:43.150 30572 31117 V HPV-ServerConnection: SSE onConnect
04-18 07:33:43.150 30572 31117 V HPV-ServerConnection: Sending pending updates...
04-18 07:33:43.150 30572 31117 V HPV-ServerConnection: Sending state update for TabletTimeStamp: 1524029613039
04-18 07:33:43.152 30572 31117 V HPV-ServerConnection: Pending updates sent
04-18 07:33:43.152 30572 31117 D HPV-ServerConnection: Actively fetching items state
```

40 seconds later HP detecting network is back:
```
04-18 07:34:12.646 30572 30572 I chromium: [INFO:CONSOLE(48)] "Error loading openHAB items... retrying in 5 seconds", source: http://rpi.fritz.box:8080/habpanel/app/services/openhab.service.js (48)
04-18 07:34:20.060 30572 30572 I chromium: [INFO:CONSOLE(35)] "Loaded 111 openHAB items", source: http://rpi.fritz.box:8080/habpanel/app/services/openhab.service.js (35)
```

State updates coming in normally from here on. Note that it takes HP 40 seconds to resume operation. There are no errors in the log in between. Even though HP states it retries in 5 seconds, there are huge gaps in between retries.
I am not sure how this works in chrome. Maybe google apps are not resource restricted by the OS?

It takes exactly 5 minutes from when you turn off the screen until the network is gone.

I will have another look at the log later when I have more time.

Can you generate a log when you do the same thing with chrome showing HP?

**Comment by Webfeger** [comment]
I turned on my tablet from standby at 18:00 and HPV Clock still displayed 12:45 
After them i close this app and start chrome webapp and turn some items to "on"
30 seconds later Tablet went to standby. I changed the state of items from other device with OpenHab App. then i turn tablet back from standby and every item was update. including time. 

I repeat it at 18:25-8:30

Its very strange why this issue exists only on my tablet.
[logcat10.zip](https://github.com/vbier/habpanelviewer/files/1924924/logcat10.zip)

**Comment by vbier** [comment]
There is not a single HPV log message in this log. I doubt it was still running at 18:00, as there would have been log messages. The only thing I can see is chrome running at 18:00. The log does not fit to what you describe you have done.

**Comment by Webfeger** [comment]
Thats strange because i added HPV App as debugging app on dev settings too.
I try to reproduce the steps again tomorrow. Till jet there are no issues detectable. Only Try to reconnect message displays more then 5 seconds

**Comment by Webfeger** [comment]
Actually it works till yet. But reconnection needs more then 20 seconds, after turn on from standby

**Comment by Webfeger** [comment]
In the meanwhile i detected some HPV issues again.
Here are the new logfile with follow Steps:
20.04 - 07:05 - Turn on tablet form standby and HPV is hanging, no reactions but time is corrct.
20.04 - 07:05 - HPV was restarted
20.04 - 07:54 - Turn on tablet from standby and HPV is hanging again, but time stay at 07:06

so i think the reason for ignoring button presses are because hpv is hanging. But the Screen-Resize icon at top right didnt work too.

[logcat11.zip](https://github.com/vbier/habpanelviewer/files/1930922/logcat11.zip)

**Comment by Webfeger** [comment]
Now it works.
I reconfigured all Battery and WLAN settings again.
Switched to 2.4. Ghz to get a higher percentage of quality,
disabled Battery optimization and disabled the app in battery optimization too. and now it works since 3 Days. I will test it again with major release of HPV.

Its strange because i didnt't find the reason why this app hangs. Now Tablet is still not able to ping after 5 minutes of standby. but HPV works well after turn on from standby.

Thanks vor extensive helps to solve this issue.

Regards, Web
